src/platform/linux/mod.rs

```diff
@@ -386,76 +390,84 @@ impl OsIpcChannel {
-        self.pollfds.push(pollfd {
-            fd: fd,
-            events: POLLIN,
-            revents: 0,
-        });
+        let io = EventedFd(&fd);
+        let fd_token = Token(fd as usize);
+        self.poll.register(&io, fd_token, Ready::readable(), PollOpt::level())
+            .expect(&format!("Could not register: {:?}", fd));
+        self.pollfds.insert(fd_token, fd);
         Ok(fd as i64)
     }
 
     pub fn select(&mut self) -> Result<Vec<OsIpcSelectionResult>,UnixError> {
         let mut selection_results = Vec::new();
-        let result = unsafe {
-            poll(self.pollfds.as_mut_ptr(), self.pollfds.len() as nfds_t, -1)
+        let sz = match self.poll.poll(&mut self.events, None) {
+            Ok(sz) if sz > 0 => sz,
+            _ => { return Err(UnixError::last()); }
         };
-        if result <= 0 {
-            return Err(UnixError::last())
-        }
 
         let mut hangups = HashSet::new();
-        for pollfd in self.pollfds.iter_mut() {
-            if (pollfd.revents & POLLIN) != 0 {
-                match recv(pollfd.fd, BlockingMode::Blocking) {
-                    Ok((data, channels, shared_memory_regions)) => {
-                        selection_results.push(OsIpcSelectionResult::DataReceived(
-                                pollfd.fd as i64,
-                                data,
-                                channels,
-                                shared_memory_regions));
-                    }
-                    Err(err) if err.channel_is_closed() => {
-                        hangups.insert(pollfd.fd);
-                        selection_results.push(OsIpcSelectionResult::ChannelClosed(
-                                    pollfd.fd as i64))
+        for i in 0..sz {
+            let evt = self.events.get(i).unwrap();
+            let evt_token = evt.token();
+            match (evt.kind().is_readable(), self.pollfds.get(&evt_token)) {
+                (true, Some(fd)) => {
+                    match recv(*fd, BlockingMode::Blocking) {
+                        Ok((data, channels, shared_memory_regions)) => {
+                            selection_results.push(OsIpcSelectionResult::DataReceived(
+                                    *fd as i64,
+                                    data,
+                                    channels,
+                                    shared_memory_regions));
+                        }
+                        Err(err) if err.channel_is_closed() => {
+                            hangups.insert(evt_token);
+                            selection_results.push(OsIpcSelectionResult::ChannelClosed(
+                                        *fd as i64))
+                        }
+                        Err(err) => return Err(err),
                     }
-                    Err(err) => return Err(err),
-                }
-                pollfd.revents = pollfd.revents & !POLLIN
+                },
+                _ => {},
```

Comment:
We could check `is_error` and `is_hup` similar to a previous version using epoll directly.